src/main/java/org/reaktivity/nukleus/http2/internal/stream/Http2ServerFactory.java

```diff
@@ -2190,13 +2248,15 @@ private void onDecodeData(
                 }
                 else
                 {
-                    if (dataLength > 0)
+                    if (payloadLength > 0)
                     {
-                        final DirectBuffer payload = http2Data.payload();
-                        exchange.doRequestData(traceId, authorization, payload, 0, payload.capacity());
+                        final int dataSent = exchange.doRequestData(traceId, authorization, payload, 0, payloadLength);
+                        progress += dataSent;
+                        dataBytesConsumed += dataSent;
```

Comment:
No need to track `dataBytesConsumed` here, it is implicit from returned `progress`, can be managed by `decodeDataPayload` instead.